src/iuse_actor.cpp

```diff
@@ -1704,3 +1704,97 @@ bool musical_instrument_actor::can_use( const player *p, const item*, bool, cons
+}
+
+void holster_actor::load( JsonObject &obj )
+{
+    holster_prompt = obj.get_string("holster_prompt", _("Holster item"));
+    holster_msg    = obj.get_string("holster_msg",    _("You holster your %s"));
+
+    max_volume = obj.get_int("max_volume");
+    min_volume = obj.get_int("min_volume", max_volume / 3);
+    max_weight = obj.get_int("max_weight", max_weight);
+    multi      = obj.get_int("multi",      multi);
+    draw_speed = obj.get_int("draw_speed", draw_speed);
+
+    allow = obj.get_string_array("allow");
+}
+
+long holster_actor::use( player *p, item *it, bool, const tripoint& ) const
+{
+    if (&p->weapon == it) {
+        p->add_msg_if_player(_("You need to unwield your %s before using it."), it->tname().c_str());
+        return 0;
+    }
+
+    int pos = 0;
+    std::vector<std::string> opts;
+
+    if ((int) it->contents.size() < multi) {
+        opts.push_back(holster_prompt);
+        pos = -1;
+    }
+
+    std::transform(it->contents.begin(), it->contents.end(), std::back_inserter(opts),
+                   [](const item& elem) { return string_format(_("Draw %s"), elem.display_name().c_str()); });
+
+    if (opts.size() > 1) {
+        pos += uimenu(false, string_format(_("Use %s"), it->tname().c_str()).c_str(), opts) - 1;
+    }
+
+    if (pos >= 0) {
+        p->wield_contents(it, pos, draw_speed);
+    } else {
+        item &obj = p->i_at(g->inv_for_filter(holster_prompt, [&](const item& e) {
+            if (e.volume() > max_volume || e.volume() < min_volume) {
+                return false;
+            }
+
+            if (max_weight > 0 && e.weight() > max_weight) {
+                return false;
+            }
+
+            return std::any_of(allow.begin(), allow.end(), [&](const std::string& str) {
+                return e.has_flag(str) || e.gun_skill() == skill_id(str);
+            });
+        }));
+
+        if (obj.is_null()) {
+            p->add_msg_if_player(_("Never mind."));
+            return 0;
+        }
+
+        // if selected item is unsuitable inform the player why not
+        if (obj.volume() > max_volume) {
+            p->add_msg_if_player(m_info, _("Your %s is too big to fit in your %s"),
+                                 obj.tname().c_str(), it->tname().c_str());
+            return 0;
+        }
+        if (obj.volume() < min_volume) {
+            p->add_msg_if_player(m_info, _("Your %s is too small to fit in your %s"),
+                                 obj.tname().c_str(), it->tname().c_str());
+            return 0;
+        }
+        if (max_weight > 0 && obj.weight() > max_weight) {
+            p->add_msg_if_player(m_info, _("Your %s is too heavy to fit in your %s"),
+                                 obj.tname().c_str(), it->tname().c_str());
+            return 0;
+        }
+        if (!std::any_of(allow.begin(), allow.end(), [&](const std::string& str) {
+                         return obj.has_flag(str) || obj.gun_skill() == skill_id(str); }))
```

Comment:
It somewhat evolved that way but I agree it would be best split, possibly as `skills` and `flags`?